rfc/src/rfcs/0008-function-contracts.md

```diff
@@ -0,0 +1,580 @@
+   *before* the call to `my_div`, limiting the generic inputs to those the
+   function is actually defined for. Postconditions (`ensures`) are inserted as
+   `kani::assert` checks *after* the call to `my_div`.
+
+   The expanded version of our harness and function is equivalent to the following:
+
+   ```rs
+   #[kani::proof]
+   fn my_div_harness() {
+     let dividend = kani::any();
+     let divisor = kani::any();
+     kani::assume(divisor != 0);
+     let result = my_div(dividend, divisor);
+     kani::assert(result <= dividend);
+   }
+   ```
+
+   This expanded harness is then verified like any other harness but also gives
+   the green light for the next step: verified stubbing.
+
+3. In the last phase the **verified** contract is ready for us to use to
+   **stub** other harnesses.
+
+   Unlike in regular stubbing Kani there to be at least one associated
+   `proof_for_contract` harness for each function to stub *and* it requires all
+   such harnesses to pass verification before attempting verification of any
+   harnesses that use it as a stub.
+
+   A possible harness that uses our `my_div` contract could be the following:
+
+   ```rs
+   #[kani::proof]
+   #[kani::stub_verified(my_div)]
+   fn use_div() {
+     let v = vec![...];
+     let some_idx = my_div(v.len() - 1, 3);
+     v[some_idx];
+   }
+   ```
+
+   To use the contract as a stub Kani must first ensure the calling context is
+   safe. It inserts a `kani::assert` for the preconditions (`requires`) before
+   the call; then it replaces the result of `my_div` with a non-deterministic
+   value. Finally the non-deterministic result is constrained by a
+   `kani::assume` of the postconditions (`ensures`).
+   
+   Mutable memory is similarly made non-deterministic, discussed later in
+   [havociing](#memory-predicates-and-havocing).
+
+   The expanded version is equivalent to this:
+  
+   ```rs
+   #[kani::proof]
+   #[kani::stub_verified(my_div)]
+   fn use_div() {
+     let v = vec![...];
+     let dividend = v.len() - 1;
+     let divisor = 3;
+     kani::assert(divisor != 0);
+     let result = kani::any();
+     kani::assume(result <= dividend);
+     let some_idx = result;
+     v[some_idx];
+   }
+   ```
+
+   Notice that this performs no actual computation for `my_div` (other than the
+   conditions) which allows us to avoid something potentially costly.
+
+Also notice that Kani was able to express both contract checking and stubbing
+with existing capabilities; the important feature is the enforcement. The
+checking is, by construction, performed **against the same condition** that is
+later used as stub, which ensures soundness (see discussion on lingering threats
+to soundness in the [future](#future-possibilities) section) and guarding against
+stubs diverging from their checks.
+
+### History Variables
+
+Kani's contract language contains additional support for reasoning about changes
```

Comment:
I would remove this first line. History variables are also relevant for parameters that are passed by value even if the function doesn't have a side effect.

I think this paragraph is more relevant for explaining assign clauses.